narwhal/worker/src/tests/worker_tests.rs

```diff
@@ -16,8 +16,109 @@ use prometheus::Registry;
 use std::time::Duration;
 use storage::NodeStorage;
 use store::rocks;
-use test_utils::{batch, temp_dir, CommitteeFixture};
-use types::{MockWorkerToPrimary, MockWorkerToWorker, TransactionsClient, WorkerToPrimaryServer};
+use test_utils::{batch, temp_dir, test_network, transaction, CommitteeFixture};
+use types::{
+    MockWorkerToPrimary, MockWorkerToWorker, TransactionsClient, WorkerBatchMessage,
+    WorkerToPrimaryServer, WorkerToWorkerClient,
+};
+
+// A test validator that rejects every transaction / batch
+#[derive(Clone)]
+struct NilTxValidator;
+impl TransactionValidator for NilTxValidator {
+    type Error = eyre::Report;
+
+    fn validate(&self, _tx: &[u8]) -> Result<(), Self::Error> {
+        eyre::bail!("Invalid transaction");
+    }
+    fn validate_batch(&self, _txs: &Batch) -> Result<(), Self::Error> {
+        eyre::bail!("Invalid batch");
+    }
+}
+
+#[tokio::test]
+async fn reject_invalid_clients_transactions() {
+    let fixture = CommitteeFixture::builder().randomize_ports(true).build();
+    let committee = fixture.committee();
+    let worker_cache = fixture.shared_worker_cache();
+
+    let worker_id = 0;
+    let my_primary = fixture.authorities().next().unwrap();
+    let myself = my_primary.worker(worker_id);
+    let name = my_primary.public_key();
+
+    let parameters = Parameters {
+        batch_size: 200, // Two transactions.
+        ..Parameters::default()
+    };
+
+    // Create a new test store.
+    let db = rocks::DBMap::<BatchDigest, Batch>::open(temp_dir(), None, Some("batches")).unwrap();
+    let store = Store::new(db);
+
+    let registry = Registry::new();
+    let metrics = initialise_metrics(&registry);
+
+    // Spawn a `Worker` instance with a reject-all validator.
+    Worker::spawn(
+        name.clone(),
+        myself.keypair(),
+        worker_id,
+        Arc::new(ArcSwap::from_pointee(committee.clone())),
+        worker_cache.clone(),
+        parameters,
+        NilTxValidator,
+        store,
+        metrics,
+    );
+
+    // Wait till other services have been able to start up
+    tokio::task::yield_now().await;
+    // Send enough transactions to create a batch.
+    let address = worker_cache
+        .load()
+        .worker(&name, &worker_id)
+        .unwrap()
+        .transactions;
+    let config = mysten_network::config::Config::new();
+    let channel = config.connect_lazy(&address).unwrap();
+    let mut client = TransactionsClient::new(channel);
+    let tx = transaction();
+    let txn = TransactionProto {
+        transaction: Bytes::from(tx.clone()),
+    };
+
+    // Check invalid transactions are rejected
+    let res = client.submit_transaction(txn).await;
+    assert!(res.is_err());
```

Comment:
nit: Could you please also check the type of error here to ensure that is the expected one and is not another error ?

Reply:
With the `NilValidator` this is using, this is somewhat moot, isn't it?